exp, log and sqrt for Transpose and Adjoint

DebadityaPal opened the pull request:
As per the guidance in JuliaLang/LinearAlgebra.jl#781, implemented exp, log and sqrt functions for Transpose and Adjoint type matrices. Added relevant tests as well.

Closes JuliaLang/LinearAlgebra.jl#781.

Keno committed:
threads: Pass env through to thread_exec

We want to start running rr on CI, but at the moment the threads
test is clearing ENV, causing it to lose the LD_PRELOAD that improves
rr performance. Fix that by instead passing through the environment
and setting the relevant JULIA_NUM_THREADS flag on top of it.

Keno committed:
FileWatching: Dump open file descriptors on failure

If the FDs we allocated are too high, we leaked fds somewhere.
On linux it's fairly easy to dump out information about all
active FDs, so do that in the failure case, so the logs can
help in debugging.

Keno committed:
Don't detach rr workers in SharedArrays tests

By default, I'm having the test suite set up to put any workers spawned
with addprocs into their own rr sessions. This doesn't work with SharedArrays,
because those workers share memory with the main process. Add a new JULIA_RR
environment variable that specifies which rr to use for the test suite and
a flag to the testsuite's addprocs command to determine whether or not to
detach the workers from the current rr session.

Keno committed:
Disable thread affinity test under rr

RR changes the affinity, so 0,1 may not be available in the affinity
mask. It's a bit complicated to fix this test and would require parsing
the existing affinity mask, so just disable it for now.

yuyichao committed:
Fix Dict key check in type inference

This has been there since #27126 (9100329d189b3d2ac6d1e03bfc7431cbe8d18c00)

Dudging by the use of `reverse_mapping[stmt_val]` below
this should either be a branch condition or an assert.

yuyichao committed:
Some `getfield` related fixes in inference

* Use the field index passed in in `lift_leaves`

    The caller has already done all the computation including bound checking.
    The `field` computed in this function is also affecting all the following iterations
    which is almost certainly wrong.

* Remove unnecessary type check on `field` in `lift_leaves` since it is always `Int`

* Move a branch disabling `return nothing` higher up

* Remove some duplicated calculation on field index in `getfield_elim_pass!`

* Fix `try_compute_fieldidx` to return `nothing` for non-`Int` `Integer` field index.

    This can cause `getfield_nothrow` to return incorrect result.
    It also gives the caller worse type info about the return value.

* Teach `getfield_nothrow` that `isbits` field cannot be undefined and getfield on such field cannot throw.

    This is already handled in `isdefined_tfunc`.

* Fix a few wrong use of `isbits` in dead branches

----

Ref #26948 (fa02d34496eb2bed9a55e4f77cb135a6b4180edc)
Ref #27126 (9100329d189b3d2ac6d1e03bfc7431cbe8d18c00)

dkarrasch reviewed (commented):
Looks good. Just one style comment, and one proposal to be discussed.

DebadityaPal commented on stdlib/LinearAlgebra/src/dense.jl:
```diff
@@ -553,6 +553,8 @@ julia> exp(A)
 """
 exp(A::StridedMatrix{<:BlasFloat}) = exp!(copy(A))
 exp(A::StridedMatrix{<:Union{Integer,Complex{<:Integer}}}) = exp!(float.(A))
+exp(A::Adjoint) = adjoint(exp(parent(A)))
+exp(A::Transpose) = transpose(exp(parent(A)))
```
In line 554,
```
exp(A::StridedMatrix{<:BlasFloat}) = exp!(copy(A))
```
This already uses the exp! function, so shifting to this format wouldn't break consistency, I think. So, should I keep the files as is or should I modify each one to use the exp! function instead.

dkarrasch replied:
Let's hear what others say. I'm inclined to keep them as they are.

mcabbott replied:
Perhaps the signature here should be `exp(A::AdjOrTransAbsMat)`, else it will catch `exp(rand(3)')`. Possibly even further restricted, for instance should this apply to `exp(view(rand(3,3), [1,3], [1,3])')` or `exp(sprand(2,2, 0.5)')`? I'm not sure.

Would vote slightly for the version which returns an ordinary matrix.

sethaxen replied:
Since `exp` is the exported function, It makes sense to me to call `exp(parent(A))`, since `exp`, not `exp!`, is more likely to be overloaded for a custom matrix type. We could also consider adding overloads for `exp!` as well, e.g.
```julia
exp!(A::Adjoint) = adjoint(exp!(parent(A)))
```
if users want to overload `exp!` for their custom matrix type. But currently `exp!` is only defined for `StridedMatrix` so this is probably not necessary.

DebadityaPal commented on stdlib/LinearAlgebra/test/dense.jl:
```diff
@@ -145,9 +145,15 @@ end
+            @test sqrt(transpose(asym))*sqrt(transpose(asym)) ≈ transpose(asym)
+            @test sqrt(adjoint(asym))*sqrt(adjoint(asym)) ≈ adjoint(asym)
+
+
```
My bad, these skipped my attention. I have fixed the issue now

vchuravy committed:
[Make] remove BINARYBUILDER_LLVM_ASSERTS and use LLVM_ASSERTIONS instead

kescobo committed:
add getindex method and tests for NamedTuple

kescobo committed:
fix stupid typos

kescobo committed:
more stupid typos

Keno committed:
Exclude `threads` test from rr tracing

nalimilan committed:
Make `findall` faster for AbstractArrays (#37177)

The `findall` fallback is quite slow when predicate is a small function
compared with generating a logical index using `broadcast` and calling `findall`
on it to compute integer indices. The gain is most visible when predicate is true
for a large proportion of entries, but it's there even when all of them are false.
The drawback of this approach is that it requires allocating a vector of `length(a)/8`
bytes whatever the number of returned indices.

imciner2 committed:
Ensure underscore is applied to trampoline symbols on OSX [#38925]

imciner2 committed:
Export jl_n_threads from the public libjulia library [#38925]

stevengj committed:
Unicode-compliant islower/uppercase (#38574)

* Unicode-compliant islower/uppercase

* don't test isletter for non-L* letters

* include titlecase in alphas test

* add news

JeffBezanson committed:
fix #38888, pessimistic sparam inference with concrete upper bound (#38899)

cmcaine committed:
Add 2-arg versions of findmax/min, argmax/min (#35316)

Defines a descending total order, `isgreater` (not exported), where unordered values like NaNs and missing are last. This makes defining min, argmin, etc, simpler and more consistent. Also adds 2-arg versions of findmax/min, argmax/min. Defines and exports the `isunordered` predicate for testing whether a value is unordered like NaN and missing. Fixes #27613. Related: #27639, #27612, #34674.

Thanks to @tkf, @StefanKarpinski and @drewrobson for their assistance with this PR.

Co-authored-by: Jameson Nash <jameson@juliacomputing.com>
Co-authored-by: Takafumi Arakaki <takafumi.a@gmail.com>

kshyatt committed:
Missing tests for arrays (#38926)

tpapp committed:
mention that InteractiveUtils is loaded automatically in interactive mode. (#38937)

eschnett committed:
Correct formatting for `sizehint!` documentation (#38939)

staticfloat committed:
[cli/trampolines]: Fix `aarch64-apple-darwin` trampoline ASM syntax

It turns out that aarch64 assembly syntax uses `;` as the comment
character, not as the statement separator.  So we need to polyfill
that.  We also set an alignment of 4 bytes (`2^2`) which is required on
Apple targets, but is also a good idea on most other aarch64 machines.

staticfloat committed:
[cli/trampolines]: Fix `i686-w64-mingw32` name mangling

Although we were dispatching to the right symbol, the name of the
re-exported symbol was lacking its leading underscore.

Keno committed:
Merge pull request #35539 from JuliaLang/kf/stdlibtests

Test suite tweaks for rr

staticfloat committed:
Define `jl_n_threads` in only `libjulia`

Without this, it was defined in both `libjulia` and
`libjulia-internal`, causing `libjulia-internal`'s updates to not be
visible.

belamenso committed:
Ensure errors in startup.jl don't block the REPL (#38261)

imciner2 committed:
Fix visibility of jl_n_threads

JL_DLLEXPORT should be used because it will automatically switch
between importing and exporting the symbol as needed.

musm committed:
Correct exponent_max to conform to IEEE definition (#38929)

* Exponent exponent_max returns actual maximum exponent, not the sentinel.


Co-authored-by: mschauer <moritzschauer@web.de>

imciner2 committed:
Allow libjulia to contain non-pointer variables

Before it was always creating the variables as a pointer
even if they are being referenced as a differnt type in the
internal library.

imciner2 committed:
Remove second declaration of jl_n_threads

The declaration has been moved to julia.h, which is already included
in this file.

DilumAluthge committed:
Support passing `auto` as the value to the `JULIA_NUM_THREADS` environment variable

aviatesk committed:
minor inferrability tweak for `manifest_deps_get` (#38954)

GunnarFarneback committed:
Repair jl_init. (#38950)

* Export jl_get_libdir and use it from jl_init to find libjulia.

staticfloat committed:
Use proper interprocedural register on aarch64

Sacha0 committed:
Carry MBedTLS patch that works around CMake 3.18.2 bug. (#38933)

MBedTLS fails to build with CMake 3.18.2, due to a bug in
CMake 3.18.2 causing failure to find python3. For more information,
please see https://github.com/ARMmbed/mbedtls/issues/3690 and
https://gitlab.kitware.com/cmake/cmake/-/issues/21204.

This pull request applies the MBedTLS patch that works around this
issue from https://github.com/ARMmbed/mbedtls/pull/3691, which
has been merged into MBedTLS's development branch but not
yet into a tagged release. The latest minor release as of that
merge was 2.25.0, so it's probably safe to conjecture that this
patch will be live as of minor release 2.26.0; it's also
possible that this patch will land on patch releases for
2.24.0 and 2.25.0, but this author doesn't know.

staticfloat committed:
Source tarballs should contain `StdlibArtifacts.toml` files (#38962)

Without this, our full source dist tarballs still try to download these.

staticfloat committed:
Adapt Windows `cglobal()` search for `libjulia-internal`

On Windows, we have a special list of libraries that we search for
default symbol resolution.  Now that we have symbols defined in
`libjulia` and then imported by `libjulia-internal`, we need to make
certain that we search `libjulia-internal` first (so that we find
non-trampoline functions first) and then `libjulia` (so that we do in
fact eventually find the symbols at all).

vtjnash committed:
build: pack checksums into fewer files by target (#38963)

Hopefully this should preserve the ability to generate these in parallel
(and keep a clean tree) and prevent conflicts when updating different
deps targets, while consolidating any related files

staticfloat committed:
[Artifacts] Note the platform we're looking for in our error message (#38960)

musm committed:
Remove stale contrib files (#38970)

belamenso committed:
Fix #23646: relpath recognizes Windows drive letters (#38259)

vtjnash committed:
Merge pull request #38938 from imciner2/im/symfix

Fix libjulia symbol trampolines, and export `jl_n_threads`

timholy committed:
Assert that _artifact_str returns a String (#38975)

I have no idea whether this is important, but I noticed this while investigating triggers for inference in a package. Feel free to close if you think this is irrelevant.

Incidentally, `_artifact_str` is a ~60ms precompile (that's just the inference, not the codegen). I see that its precompile was commented out in 924484f2f00; just mentioning it in case there's an opportunity to squeeze out a bit more latency-reduction.

vtjnash committed:
codegen: use new freeze intrinsics to avoid poison UB (#38977)

LLVM would likely be required to insert this anyways to make hoisting
legal (converting branches to and/or), so might as well add it
ourselves, so that users can legally write that code directly.

Keno committed:
Refactor JL_GC_PUSHFRAME to internalize layout assumptions (#38982)

Right now callers of JL_GC_PUSHFRAME need to be aware that the
roots start three pointers after the end of the struct. That
seems like an implementation detail of the macro, so refactor the
macro to also take the assignment location for the locals instead.
In the process fix an error where the gc-analyzer version of this
code assumed a different layout.

Commit:
Document the opposites of any and all (#38985)

Commit:
Use occursin instead of findfirst(...) === nothing (#38990)

kshyatt committed:
Tests for push and pushfirst for abstractarray (#38991)

musm committed:
Document __module and __source (#38987)

Co-authored-by: June-6th <2660994113@qq.com>

simeonschaub committed:
use parent type in similar(::ReinterpretedArray) (#38965)

We already do this for `SubArray`, `PermutedDimsArray`, and `ReshapedArray`, so I think this makes sense here. This also adds tests for `similar(::ReshapedArray)`, since I didn't notice that was actually defined before.

ref https://github.com/JuliaGPU/CUDA.jl/issues/602

sostock committed:
Correct handling of non-Booleans in eval_test (#38695)

ronisbr committed:
Arg. to avoid escaping backslahes (escape_string) (#38597)

* Arg. to avoid escaping backslahes (escape_string)

This commit adds an argument to `escape_string` to assume that all
backslashes (`\\`) are already escaped. Hence, `\\cdot` becomes `\\cdot`
instead of `\\\\cdot`.

The default value to this argument is `false`. Hence, this does not
break any existing code since `escape_string` behaves exactly the same
as before if no additional option is passed.

Closes #34042

* Replace `bsescaped` arg. with `keep` keyword

* Update NEWS.md

* Fix doc tests

* Add compat annotation

Co-authored-by: Ronan Arraes Jardim Chagas <ronan.arraes@MacBook-Pro-de-Ronan.local>

stevengj committed:
extend iseven/isodd to Number (#38976)

* extend iseven/isodd to Number

* add PR num to NEWS

kshyatt committed:
Missing tests for accumulate (#38943)

vchuravy committed:
Construct constant LLVMPtr correctly (#38958)

Co-authored-by: Julian Samaroo <jpsamaroo@jpsamaroo.me>

Co-authored-by: Julian Samaroo <jpsamaroo@jpsamaroo.me>

Volker-Weissmann committed:
fixed a bug in the build system (#38997)

vchuravy committed:
Fix trampoline on PPC (#38980)

* [Make] Normalize ppc64le to powerpc64le

* [CLI] Fix trampoline on PowerPC

1. Add global entry
2. Don't store `r2` on parent frame
3. Conservative load from toc

* remove jl_compile_extern_c from exported funcs

aviatesk committed:
define at-invoke macro: (#38438)

- provides easier syntax to call `Core.invoke`, e.g. `@invoke 
f(a::Integer)` will be expanded into `invoke(f, Tuple{Integer}, a)`
- when type annotation is omitted, the argument type is specified as 
`Any`

Built on top of #37971

Keno committed:
Remote restriction on documenting empty baremodules (#38998)

Replaces and closes #13109. The underlying bug here appears to have
been fixed and the restriction here was never triggered since lowering
now always inserts LineNumberNodes, but we might as well remove the
restriction since the bug is fixed.

omus committed:
Add reference to `indent` function (#38940)

* Add reference to indent function

* Update base/strings/io.jl

Co-authored-by: Steven G. Johnson <stevenj@alum.mit.edu>

Co-authored-by: Keno Fischer <keno@juliacomputing.com>
Co-authored-by: Steven G. Johnson <stevenj@alum.mit.edu>

Volker-Weissmann committed:
makefile: better debug comment in output (#39002)

stevengj committed:
move `escape_string` NEWS (from #38597) to the correct section [ci skip]

stevengj committed:
move NEWS from #38675 to the correct section [ci skip]

timholy committed:
Update the AST docs on `nothing` `:method` expressions (#38496)

Fixes #38451

staticfloat committed:
Improve consistency of SONAME usage, fix debug installation

SONAME and RPATH inconsistency resulted in `julia`/`julia-debug`
occasionally being unable to find `libjulia.so`, due to
`libjulia-internal` not having the right RPATH set (it was set to
`$ORIGIN:$ORIGIN/julia`, when it now needs to be set to
`$ORIGIN:$ORIGIN/..`).  Use `patchelf` to address this during `make
install`.

We also want to use the SONAME of e.g. `libjulia` as much as possible,
so as to avoid searching the filesystem for libraries that are already
opened.

Also fix a few bugs in building `julia-debug`.

staticfloat committed:
Use `strcmp()` to compare `f_lib` and `JL_LIBJULIA_DL_LIBNAME` (#39005)

This future-proofs us from someone trying lookup a symbol from
`libjulia.dll`, even though that should be quite rare, it may happen.

staticfloat committed:
[loader] Must invalidate `.o` files when `VERSION` changes (#38800)

This doesn't happen that often, (only when `VERSION` changes) but we
need to be sensitive to the path to `libjuliainternal.so` changing.

staticfloat committed:
Fix lack of `JL_MAJOR_SHLIB_EXT` on Windows

simeonschaub committed:
add missing newline to UnionAll Vararg warning (#39010)

musm committed:
win: Set correct folder permissions after folder creation (#38942)

libuv's mkdir does not implement the mode argument on Windows, as a
result the folder mode is never passed on the Windows platform.

Forcibly change the folder permissions to agree with the mode passed
into `mkdir`.

rapus95 committed:
keys(::Generator) pass through (#34678)

* keys(::Generator) pass-through

Let keys pass through to allow lightweigth/lazy find* and arg* functions

* Update test/functional.jl

Co-authored-by: Matt Bauman <mbauman@gmail.com>

Co-authored-by: Matt Bauman <mbauman@gmail.com>
Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

vtjnash committed:
Merge pull request #39009 from JuliaLang/sf/dlopen_libjulia

Improve consistency of SONAME usage, fix debug installation

kescobo committed:
test suggestions from jw3126

musm committed:
Revert "win: Set correct folder permissions after folder creation (#38942)" (#39014)

This reverts commit 556cee7f0971407eaf547e291d66041e16293b6b.

nalimilan committed:
Improve docs about let blocks (#39003)

This section was probably written before the distinction between
hard and soft scopes was added.

fingolfin committed:
Remove outdated references to Travis (#39011)

musm committed:
filesystem: code cleanup in request calls (#39012)

The canonical way we allocate the libuv requests in the filesystem
API is through malloc/free calls. Clean up the code to be more DRY.

vtjnash committed:
compiler: corrections to Conditional lattice

Try harder to equate Const(bool) with the Conditional represenation of the same

vtjnash committed:
inference: remove dead check for NOT_FOUND

Missed in dcc0696971a100dbc63f90b8789632631dd1bfef cleanup

aviatesk committed:
add news for `at-invokelatest` and `at-invoke` (#39000)

Keno committed:
Fix insert_node_here! if called at the end of the IR (#39021)

This method was throwing an error when called after the compaction
had processes the last instruction in the IR (with reverse affinity).
This currently can't, because we never use this method after compacting
a terminal instruction (inlining uses it to split BBs), but can after
the new code in #37849, so fix the bug now. Forward port from #37849.

Keno committed:
Change arguments to (internal) `show_tuple_as_call` to kwargs (#39020)

Forward port from #37849 to reduce the number of unrelated changes
in that PR. The argument list here was simply getting unwieldy.
The new `hasfirst` argument isn't yet used in this PR, but will
be used in #37849.

JeffBezanson committed:
fix IR showing not to show TypedSlot type twice (#38961)

* fix IR showing not to show TypedSlot type twice

* Update base/compiler/ssair/show.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Co-authored-by: Keno Fischer <keno@juliacomputing.com>
Co-authored-by: Jameson Nash <vtjnash@gmail.com>

dkarrasch committed:
Improve performance for dot(AdjTrans,AdjTrans) (#39004)

* Improve performance for dot(AdjTrans,AdjTrans)

* generalize to Adjoint/Transpose

KristofferC committed:
remove unnecessary ::Union{} type asserts (#35817)

* remove unnecessary ::Union{} type asserts

Keno committed:
Document destructuring syntax in anonymous function argument (#39024)

When destructuring a single tuple in an anonymous function argument,
you have to include an extra comma to force a 1-tuple. Otherwise ((x,y))
collapses to (x,y). This is not the case for named functions.

Co-authored-by: Graham Smith <grahamas@gmail.com>

Keno committed:
Merge pull request #39036 from JuliaLang/kf/compilerroundup

Round up and rebase approved compiler PRs

staticfloat committed:
[libuv] Bump to include `DELETE` win ACL patch (#39038)

This bump to LibUV adds an important permission to files which should
allow them to be deleted more reliably when created in "system"-like
locations such as `C:\`.

sostock committed:
Avoid StepRangeLen eltype widening (#38979)

simeonschaub committed:
specialize on mapfoldl(::Type, ...) (#39019)

Before:
```julia
julia> using BenchmarkTools

julia> b = rand(Bool, 10000);

julia> @btime mapfoldl(Int, +, b; init=0)
  1.136 ms (9025 allocations: 141.02 KiB)
5057
```

After:
```julia
julia> @btime mapfoldl(Int, +, b; init=0)
  1.094 μs (1 allocation: 16 bytes)
5057
```

vchuravy committed:
mark integer parameters as sext/zext

vchuravy committed:
silence unused variable warning

staticfloat committed:
[loader/libjulia]: Call `setup_stdio()` when loading `libjulia`

This allows libjulia's initialization routines to properly output to
`stdout` in embedding scenarios.

staticfloat committed:
[cli/trampolines] Fix section directive for windows trampolines

By naming the section `text` instead of `.text` these sections were
mapped into memory without the executable bit set, causing segfaults
when attempting to use any trampolines on Windows.

staticfloat committed:
[cli/loader]: Don't allow initialization to run more than once

There are multiple ways of fixing this, but let's just not let
trampolines get re-set after they've already been set.

kshyatt committed:
Test for checkout_head (#39034)

* Test for checkout_head

* Remember to actually close the file

* And actually force the checkout

Keno committed:
Merge pull request #39052 from JuliaLang/sf/win_embedding_fixes

Various fixes to trampolines for embedding on Windows

Keno committed:
Improve copyto! for short heterogeneous tuples (#39035)

There was a complaint on Twitter [1], that Julia is slower than
Ruby(!) at allocating short heterogeneous arrays. The code looks
something like this:

```
f() = for i in 1:100000; b=[1,2.0]; end;
```

Of course, this benchmark is silly because the arrays are
unused, so the entire code should be dropped once we have
further improvements to our lifetime analysis, which will
change the runtime 0. However, since we're not quite able
to do that yet, it is somewhat unexpected that we should
be slower than Ruby here. Admittedly our arrays are tuned
for larger sizes, but still. Upon further investigation,
it turns out that the issue is that we're allocating extra
boxed objects when pulling the fields out of the literal,
since we don't have any specialization for `copyto!` from
tuples, falling back to the generic `copyto!` method, which
is type unstable in the presence of heterogeneous tuples.

Since this code is reachable from surface syntax, I think
it's worth putting in a couple of specializations to
accelerate the array construction, which is what this PR
does (up to tuples of length 10 to avoid excessive
codegen). Since the type-instability issue applies whether
the value is used or not, this should still be useful even
if we're able to prove that the allocations can be dropped
entirely.

Before:
```
julia> @benchmark f()
BenchmarkTools.Trial:
  memory estimate:  21.36 MiB
  allocs estimate:  600000
  --------------
  minimum time:     14.928 ms (0.00% GC)
  median time:      16.297 ms (0.00% GC)
  mean time:        16.323 ms (3.84% GC)
  maximum time:     26.042 ms (0.00% GC)
  --------------
  samples:          307
  evals/sample:     1
```

After:
```
julia> @benchmark f()
BenchmarkTools.Trial:
  memory estimate:  9.16 MiB
  allocs estimate:  100000
  --------------
  minimum time:     2.117 ms (0.00% GC)
  median time:      2.276 ms (0.00% GC)
  mean time:        2.455 ms (7.62% GC)
  maximum time:     4.638 ms (28.99% GC)
  --------------
  samples:          2036
  evals/sample:     1
```

Which puts us about on par with Ruby on my machine (not entirely
unsurprising, since this is essentially an allocation benchmark
at this point and both allocators are written in C).

Keno committed:
Add `ismutabletype` (#39037)

Determines whether a type was declared using `mutable struct`.
Naming follows from the `ismutable` query we already have, analogous
to `isbits`/`isbitstype`. Replaces #18168.

shikhargoswami committed:
Written tests for typemin/typemax (#39057)

* Written tests for typemin/typemax

* Added test w.r.t Int128

fingolfin committed:
Fix jl_new_task to avoid allocating stack twice

If the user requested dedicated stack of a certain size, then `jl_new_task`
immediately allocates the stack by invoking `jl_alloc_fiber`.

However, it also later nulls `t->stkbuf`. As a result, `ctx_switch` invokes
`jl_alloc_fiber` *again* to allocate another stack.

aviatesk committed:
minor typing improvements for `(frame::InferenceState).stmt_edges`

gokyo committed:
LICENSE: copyright end date 2019 -> 2021 (#39084)

sethaxen commented:
Just a note, but `f(adjoint(x)) == adjoint(f(x))` is a general property of matrix functions whose power series representations have real coefficients. So these overloads could also be added for the matrix trigonometric and hyperbolic functions, as well as their inverses. Although, it seems that many of these already work, and the rest that I checked worked as soon as `exp` was implemented. Perhaps tests could be added for these to protect from future regressions.

vchuravy committed:
add patches for PPC

oxinabox committed:
Remove outdated TODO about string encodings (#39077)

Keno committed:
Small drive-by cleanup: Make constant explicit (#39056)

This code is a bit of a leftover from when we still had `nfields(::Type)`,
but at this point it was always just returning the number of fields
of `DataType`. I was a bit confused by this code, so I thought
just using `jl_datatype_t` here explicitly makes clear that it does
actually just return said constant.

dlfivefifty committed:
Fixes for non-Int based lengths (#37741)

Sacha0 committed:
Merge pull request #39080 from aviatesk/typingedges

minor typing improvements for `(frame::InferenceState).stmt_edges`

matbesancon committed:
fix inf rational (#39063)

vchuravy committed:
don't dereference Expected

staticfloat committed:
[Artifacts]: Download to temporary directories, not `stdlib`

We can't assume that `stdlib` is in a writable location; always download
temporary testing files to `$TMPDIR`.

mortenpi committed:
Bump Documenter to 0.26.1 (#39097)

staticfloat committed:
[REPL]: Don't write to the user's home directory

Instead of writing out to actual `$HOME`, create a temporary directory
and set that as `$HOME`/`$USERPROFILE`.

staticfloat committed:
[REPL] Do not require test to run within `/tmp`

mroavi committed:
fix typo in metaprogramming.md (#39095)

Fix typo

vchuravy committed:
[JITLayers] Consume errors on address lookup

musm committed:
filesystem: use libuv for rmdir directly (#39013)

imciner2 committed:
Backport LLVM patch to fix LLVM build on GCC 11

GCC 11 has changed header dependencies again, so the build of LLVM
fails because <limits> wasn't being included when it should be. This
is already fixed upstream in LLVM 12.

Keno committed:
Merge pull request #39101 from JuliaLang/sf/sandbox_fixes

Fixes so that test suite can run in more tightly sandboxed environment

IanButterworth committed:
Measure compile time only when using time macros pt.2: handling when code under test throws and make compile timing thread-local (#38915)

* ensure compile timing disables in time & timev macros

* make comp measurement switching threadsafe

simeonschaub committed:
fix Meta.partially_inline! again (#39112)

forgot to take `statement_offset` into account in #38519

musm committed:
Test for Windows delete permissions in system folders (#39078)

aviatesk committed:
inference: minor typing improvements for `InferenceState.stmt_types` (#39096)

aviatesk committed:
inference: minor code quality improvements

josemanuel22 committed:
Add a warning that HTML should not be used #38909 (#39072)

* Add a warning that HTML should not be used #38909

IanButterworth committed:
Remove try from at-time and close compile timer during throw (#39133)

* remove try from at-time and close compile timer during throw

* add scope tests for at-time and aat-timev

vtjnash committed:
Revert "Improve copyto! for short heterogeneous tuples (#39035)" (#39128)

This reverts commit b00e9f0bac05de913d5b869739bcc93ac259c9f7, which has a 1/64 chance of failing tests.

maleadt committed:
Fix alloc opt pass on LLVM 11+. (#39119)

JeffBezanson committed:
fix part of #38936, getfield elim handling union of tuples (#39107)

This fixes the regression since 1.5

staticfloat committed:
`addenv()`, by default, should inherit from `ENV` (#39100)

The spirit of `addenv()` is to be non-destructive, however because a
`Cmd` object with an `env` member that is set to `nothing` (the default)
inherts from the current environment when it is run.  This means that
the following, rather surprising behavior holds true on current master:

```
julia> ENV["FOO"] = "foo"
       run(`/bin/bash -c "echo \$FOO \$BAR"`)
       run(addenv(`/bin/bash -c "echo \$FOO \$BAR"`, "BAR" => "bar"))
foo
bar
```

This PR adds an `inherit` flag to `addenv()` to allow keeping this
behavior if it is actually desired (this might make sense if you are
constructing `.env` blocks over a series of `addenv()` calls and you
don't want to have to special-case your first operation as a `setenv()`
rather than an `addenv()`.

vtjnash committed:
remove Core._apply Builtin (#39115)

We can emulate this deprecated function, until we delete it in v2.
This in turn revealed some missing error checks, which we also add.

Fixes #39113

JeffBezanson committed:
fix #39117, overly-specific type from `default_eltype` (#39130)

vchuravy committed:
Merge pull request #39051 from JuliaLang/vc/sext

Annotate function arguments with sext/zext

DebadityaPal committed:
added automatic keyword assignment support to test macro (#38270)

* added automatic keyword assignment support to @test macro

* added some tests for test macro using atol keyword

* x = a.x syntax support added

* Update stdlib/Test/src/Test.jl

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

JeffBezanson committed:
fix some compiler warnings (#39142)

- unused jl_iterate_func
- cast type of realloc in jl_init_threading

JeffBezanson committed:
make methods of `alloc_buf_hook` return the same types (#39141)

vtjnash committed:
Core printing: improve show of methods (#39135)

This adds their environment, so that their TypeVars will print correctly.

simeonschaub committed:
fix count(::BitArray; dims) (#39149)

Not sure why this worked before #37461, perhaps #9498?

vtjnash committed:
gf: avoid adding cache entries wider than the original method (#39140)

Sometimes we want to widen the compilation signature, but then end up
with something which does not fit the original pattern. This then can
cause problems later, when we try to use the Method (from the cache),
but discover it does not actually match the call.

Fixes #38999

marius311 committed:
fix logdet(Diagonal{<:Real}) with negative entries (#39158)

aviatesk committed:
minor code quality improvements (#39159)

mkitti committed:
Remove start_repl_server code from REPL module (#39162)

`REPL.start_repl_server(port::Int)` uses outdated API and no longer functions. It should be removed.

```julia
julia> import REPL

julia> REPL.start_repl_server(8080)
ERROR: MethodError: no method matching listen(::REPL.var"#80#81", ::Int64)
```

Keno committed:
subtype: rm dead code

This code has been dead since #38136. Clean it up some.

quildtide committed:
Replace broken link in distributing.md

Keno committed:
subtype: Clean up tuple vararg path some more

Now that we no longer need to introduce tvars, we can get rid
of the closure env and reorganize the code a bit.

DilumAluthge committed:
Add the public `Dates.periods` function for getting the `Vector` of `Period`s that comprise a `CompoundPeriod`

Keno committed:
Turn check that's always true into an assert

bicycle1885 committed:
make lowrankupdate/downdate more permissive (#39150)

* make lowrankupdate/downdate more permissive

* add tests for abstract vectors

vchuravy committed:
Merge pull request #39166 from Thalassocracy/fix-documentation

Replace broken link in distributing.md

vtjnash committed:
inference: fix vararg normalization in rewrap (#39134)

Fixes #39082

DilumAluthge committed:
Apply suggestions from code review

DilumAluthge committed:
Fix whitespace

quinnj committed:
Merge pull request #39169 from JuliaLang/dpa/compoundperiod-periods

Add the public `Dates.periods` function for getting the `Vector` of `Period`s that comprise a `CompoundPeriod`

sostock committed:
Add compat note in Dates.periods docstring (#39189)

vtjnash committed:
lowering: remove incorrect comprehension eta reduction (#39139)

But keep it for `[f(_) for _ in x]`, since that would now triggers a
deprecation error (which was previously incorrectly missing for _).

Fixes #18621

vtjnash committed:
improve redirect_std* functions (#39132)

Handles Pipe now as an argument (as documented), and reduces use of
metaprogramming by introducing a name for the common functionality.

Fixes #13519

DilumAluthge committed:
Docs and news for #38952 (#39032)

* Update the docs for the `JULIA_NUM_THREADS` environment variable
* Add NEWS for #38952

Keno committed:
Merge pull request #39174 from JuliaLang/kf/varargsubtype

Clean up subtyping of varargs

staticfloat committed:
Fix `p7zip_jll` environment setup (#39155)

OvidiusCicero committed:
update latest release in README (#39197)

README should advise to checkout last bugfix release 1.5.3 instead of first full 1.5.0

simeonschaub committed:
improve inferrability of tuple slicing (#39074)

* RFC: improve inferrability of tuple slicing

This makes slicing tuples type stable in a lot more cases. This now only calls `ntuple` if the resulting tuple has no more than 10 elements, as just relying on the fallback in `ntuple` did have some compile-time overhead in my artificial benchmarks. I also widened the signature to `AbstractUnitRange`, since I don't see why we shouldn't have this for ranges like `OneTo` as well.

* add check for offset ranges and more tests

* add guard if OffsetArrays is already imported

Co-authored-by: Matt Bauman <mbauman@juliacomputing.com>

* import OffsetArrays correctly

Co-authored-by: Matt Bauman <mbauman@juliacomputing.com>

jw3126 committed:
Improve range: refactor, support start as an optional kwarg, clearer docs and error messages (#38041)

Mathematically a range is uniquely determined by three out of four of start, step, stop, length.
Furthermore if one assumes step=1 any combination of two others macthematically suffices to specify a range.

With this PR the range function reflects this. Any combination of three (two non step) arguments will be accepted.

Co-authored-by: Michael Abbott <32575566+mcabbott@users.noreply.github.com>
Co-authored-by: Johnny Chen <johnnychen94@hotmail.com>
Co-authored-by: Mark Kittisopikul <mkitti@users.noreply.github.com>
Co-authored-by: Matt Bauman <mbauman@juliacomputing.com>

timholy committed:
Test & CoreLogging: improve inference & (no)specialization (#39177)

This adjusts specialization in both Base.CoreLogging and Test,
and transitions to concrete types in some Test structs.
It also fixes a couple of inference problems.

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

dkarrasch committed:
[WIP] Speed up dense-sparse matmul (#38876)

* Speed up dense-sparse matmul

* add one at-simd, minor edits

* improve A_mul_Bq for dense-sparse

* revert ineffective changes

* shift at-inbounds annotation

martinholters committed:
Let tmerge form a Union more often (#27843)

Let tmerge decide whether to directly form a Union based on unioncomplexity, not
concreteness.

vchuravy committed:
move LLVM assert subversion to llvm.mk

vchuravy committed:
update LLVM binaries

vchuravy committed:
use BINARYBUILDER_LLVM_ASSERTS for downloading assert builds

vtjnash committed:
inference: correction to ifelse Conditional lattice

Rename typeassert_type_instance to tjoin (aka typeintersect). Also,
since the ifelse value here might not be in the regular type lattice, we
need to use the extended lattice for this evaluation.

timholy committed:
Add some costly precompiles (#39179)

These are things that have shown up in analysis of various packages
by SnoopCompile. Either they are particularly costly, or in the
case of `Test`, just plain annoying to have them show up again
and again.

fingolfin committed:
Fix a compiler warning

kescobo committed:
remove comments from tests

omus committed:
Add libldl to SuiteSparse libs (#39211)

omus committed:
Copy libgomp as part of CSL build (#39212)

vtjnash committed:
Merge pull request #39221 from JuliaLang/jn/conditional-bugs

fix some issues with inference Conditionals

kescobo committed:
add compat notice and NEWS

omus committed:
Fix spawn test when env `BAR` is set (#39214)

omus committed:
MbedTLS source build fixes (#39131)

fingolfin committed:
Remove unused jl_get_ptls_states_wrapper, update comments (#39066)

Since jl_get_ptls_states_wrapper is static, it can't be called by outside code.

omus committed:
Remove csl and suitesparse from DEPS_LIBS_STAGED (#39233)

JeffBezanson committed:
fix return type for `jl_uncompress_argnames` (#39215)

vchuravy committed:
Merge pull request #39227 from fingolfin/mh/fix-warning

Fix a compiler warning

vchuravy committed:
Merge pull request #39094 from JuliaLang/vc/ppc_llvmpatches

add patches for PPC

quinnj committed:
Merge pull request #38878 from kescobo/getindex-nt

Add getindex methods for NamedTuple

omus committed:
Add distclean-csl rule (#39234)

simeonschaub committed:
allow running doctests with Revise (#38258)

Co-authored-by: Max Horn <max@quendi.de>

Co-authored-by: Max Horn <max@quendi.de>

dkarrasch committed:
Allow SVD of vectors (#39087)

rikhuijzer committed:
[doc]: correct docstring for strip (#39231)

aviatesk committed:
inference: yet more fix for `ifelse` lattice (#39247)

aviatesk committed:
inference: improve code qualities (#39250)

JeffBezanson committed:
fix #39218, bug in subtyping fast path for tuples with repeated elements (#39237)

staticfloat committed:
[Artifacts]: Support `using Pkg.Artifacts` with LazyArtifacts (#39210)

Just as we wanted to support users that used `using Pkg`, we should
support users that use `using Pkg.Artifacts` with their lazy artifacts.
This PR merely extends support to those users and adds regression tests.

staticfloat committed:
Bump `Pkg` to `9b78818f09ea7e938bbb514f2056c6241714716f`

This Pkg commit contains necessary improvements for BB on v1.6+

staticfloat committed:
Fix incorrect `PKG_BRANCH` (#39257)

vchuravy committed:
[LLVM] Adjust patch list for LLVM 11.0.1

vchuravy committed:
Merge pull request #39255 from JuliaLang/vc/assertions

Fix LLVM assertions

timholy committed:
Fix invalidations for `show(::MethodInstance)` and `detect_ambiguities` (#39253)

timholy committed:
Fix binaryplatforms `union` invalidation (#39251)

DataStructures' `union(s::DataStructures.SparseIntSet, ns)` yields
a bad invalidation.

vtjnash committed:
staticdata: fix unreferenced data assertion (#39246)

Need to sequence operations correctly here. We need to first finish
marking referenced data before we start deleting unreferenced data. This
otherwise can lead to an assertion failure in some situations.

Refs #37650

staticfloat committed:
Revert "Bump `Pkg` to `9b78818f09ea7e938bbb514f2056c6241714716f`" (#39263)

vtjnash committed:
codegen: mostly fix uses of undefined PhiNodes (#39236)

Unlike the rest of codegen, we permit PhiNodes to legally return
undefined values, so we need to carefully skip those.

Fixes #39232

JeffBezanson committed:
fix #39259, leave `Main.ARGS` unresolved on startup (#39262)

timholy committed:
Prevent TOML invalidation by pirates of `T[elements...]` (#39252)

StaticArrays (perhaps among others) semi-pirates this method,
although they've worked hard to make it pretty innocuous.
Still, there are times when it invalidates the TOML-reading,
so best to protect this.

staticfloat committed:
Bump `Pkg` to latest tip of `Pkg.jl#master` (#39264)

* Bump `Pkg` to latest tip of `Pkg.jl#master`

* add a docmeta to Pkg

Forward-porting
https://github.com/JuliaLang/julia/commit/20c94f9d6e1143eb0fe0fd96df880c1be67a75c9

staticfloat committed:
Fix macos codesign workflow (#39152)

We changed the permissions on executable files which caused this `find`
to not find any of the actual binaries that we need to sign.  Change it
to instead find anything with any executable permissions set, rather
than ones with exactly the permissions `0755`

Keno committed:
Close uv_idle handles when we're done with them (#39268)

uv_idle handles are a bit dangerous because the essentially prevent
the uv loop from going to sleep, so if there's no work to be done
on the julia side, it just turns into a busy loop. We start such
an idle callback in the threads.jl test. We make sure it goes out
of scope, so the GC will eventually close it, but in the meantime
they keep the loop spinning. Now, unfortunately, the next test in
line is the Distributed test, which runs everything in a subprocess,
so it never builds up enough memory pressure to actually run the
GC, so the idle handles never get closed. This isn't too big a
deal, as the worst thing that happens is that it hogs one CPU
core on CI while the threads test is running, but we don't have
enough parallelism there anyway, so I don't expect a meaningful
impact on CI. It does however blow up our rr traces and apparently
causes significantly problems when trying to replay them on pernosco.
This is easy to fix by just closing these handles once we're done
with them.

sijow committed:
Fix LaTeX completions of Greek variants, add \frakI and \frakR (#39148)

* Fix LaTeX completions of Greek variants, add \frakI and \frakR

* Remove \Join completion

It is redundant with \join

* Add canonical symbol->LaTeX mapping for ambiguous cases

A test checks that all ambiguous cases are handled.

* Replace \dots suggestion with \ldots

fingolfin committed:
Update manual: Vararg{T,N} is not a type anymore (#39239)

staticfloat committed:
Use arch-independent magic number as start seed for Preferences hash (#39274)

`Preferences.jl` is currently broken on 32-bit because hashing natively
uses `UInt32`'s instead of `UInt64`'s.  We allow `Preferences.jl` to
polymorph to whichever it requires here, while eliminating a confusing
large constant and simply starting from zero.

vchuravy committed:
[CCall] Don't ZExt half

mkitti committed:
Prepend build_bindir to LLVM_CONFIG_PATH_FIX rather than append (#39275)

vchuravy committed:
[LLVM] Upgrade LLVM to 11.0.1

fredrikekre committed:
quote bindings with spaces in doc search (#39273)

baggepinnen committed:
quote task_local_storage in docstrings (#39306)

Without the quoting, several terminals render the string as 
```
help?> Test.push_testset
  push_testset(ts::AbstractTestSet)

  Adds the test set to the tasklocalstorage.
```
with the word `local` being underlined.

vtjnash committed:
aotcompile: avoid cache lookup when disallowed (#39265)

Fix #38548

omus committed:
Fix tests sensitive to `--color=yes` (#39286)

omus committed:
Fix `Base.runtests("loading")` working directory (#39238)

vchuravy committed:
Add environment flag to pass extra exeflags (#39300)

Allows one to pass extra flags to test, such as forcing a specific
architecture to flush out codegen bugs.

I have been using it to test LLVM for Power8 on Power9
```
JULIA_TEST_EXTRA_EXEFLAGS=-Cpwr8
```

vtjnash committed:
Revert "compiler: corrections to Conditional lattice"

This reverts commit e1859a1c25aad42318777438dbedac485c40a7ce.
I was wrong, and lattices are hard.

vchuravy committed:
Merge pull request #39260 from JuliaLang/vc/llvm11.0.1

Update LLVM to 11.0.1

timholy committed:
Force-specialize on `T` in `cat_similar` (#39292)

These methods are tiny (quick to compile), call methods
that force-specialize on `T`, and are called by methods
that force-specialize on `T`. Consequently, there does not
seem to be any good reason to lose inferrability in these methods.

maleadt committed:
Only construct legal int types during alloc opt.

When promoting heap to stack allocations, make sure we only emit legal
integer allocas by checking with the module's datalayout. X86 doesn't
seem to care, but back-ends like SPIR-V don't know how to handle
arbitrarily-sized integers.

Fixes JuliaGPU/oneAPI.jl#55

maleadt committed:
Adapt the alloc-opt tests.

IanButterworth committed:
bump Pkg to fe759b2449b3ace291b6a5e43bd7ecff1767ff91 (#39288)

JeffBezanson committed:
improve performance of disabling finalizers in locks (#39153)

helps #38947

maleadt committed:
Merge pull request #39315 from JuliaLang/tb/legal_alloc_opt

Only create legal int types during alloc opt

timholy committed:
Preserve input type for unaliascopy(::ReinterpretArray) (#39316)

timholy committed:
Improve inferability of shape::Dims for cat (#39294)

`cat` is often called with Varargs or heterogenous inputs,
and inference almost always fails. Even when all the arrays
are of the same type, if the number of varargs isn't known
inference typically fails. The culprit is probably #36454.

This reduces the number of failures considerably, by avoiding
creation of vararg length tuples in the shape-inference pipeline.

henriquebecker91 committed:
`include` only works on global scope (#39317)

Just changed the ending of a phrase:

"`include` behaves as if the contents of the source file were evaluated in its place."

to

"`include` behaves as if the contents of the source file were evaluated in the global scope of the including module."

because I have already answered more than one person in Discourse trying to load variable definitions and code pieces from another file to inside a function scope (instead the global scope).

KristofferC committed:
Rebase: Performance regression of scalar randn() between Julia 1.4 and 1.5  (#39319)

* definition of randn for scalars reverted to 1.4

* Added comments why this change is done

Co-authored-by: Benjamin Lungwitz <52384612+lungben@users.noreply.github.com>

timholy committed:
Use lispy tuples in cat (fixes #21673) (#39314)

The `cat` pipeline has long had poor inferrability.
Together with #39292 and #39294, this should basically put an
end to that problem.

Together, at least in simple cases these make the performance
of `cat` essentially equivalent to the manual version.
In other words, the `test1` and `test2` of #21673 benchmark
very similarly.

DebadityaPal commented:
@dkarrasch, should I go forward with any particular format?

JeffBezanson committed:
make `extrema` with dims inferrable (#39321)

fixes #39281

dkarrasch commented:
Thanks for coming back to this. From the comments above, I'd suggest to change the signatures to
```julia
exp(A::Adjoint{<:Any,<:AbstractMatrix}) = adjoint(exp(parent(A)))
```
(and similar for `Transpose`) to allow other packages to benefit from this extension for their custom matrix types. I'd argue that strange types like in https://github.com/JuliaLang/julia/pull/38371#discussion_r521288923 will simply fail at the call of `exp` just like their non-adjoint counterparts, so I don't think we need to be overly restrictive here (e.g., by restricting to `Adjoint{<:Any,<:StridedMatrix}` or so). It would be great to include tests for the functions listed in https://github.com/JuliaLang/julia/pull/38371#issuecomment-753892358 and see what "just works" and what doesn't.

vtjnash committed:
inference: make Limited tracking part of the type lattice (#39116)

This helps refine our knowledge of the `[limited]` flag setting, which
previously would always exclude a result from the cache when hitting a
cycle. However, we really only need to exclude a result if the result
might be dependent on that flag setting. That makes this formally part
of the lattice, though can be annoying to work with yet another wrapper,
so we try to add/remove it late/early to propagate it when necessary.

JeffBezanson committed:
fix failing `enable_finalizers` test

barucden committed:
Docs: add note about `IOBuffer` to `countlines` and `eachline` (#39331)

pbayer committed:
Update threadingconstructs.jl (#39309)

I try to make clearer that the task **is scheduled** to run on an available thread. This makes an important difference since between the creation of the task and its allocation to a thread may be a time delay.

see: [discussion on Discourse](https://discourse.julialang.org/t/looking-for-code-to-solve-surely-common-embarrassing-parallelism-multithreading-use-case/53527/9)

sostock committed:
Avoid type widening in range_stop_length (#39283)

sostock committed:
Simplify range_start_stop_length (#39272)

Keno committed:
OpaqueClosure - Runtime & Codegen only (#39023)

This is #37849 modulo inference/optimization support.
This is self-contained and the functionality tests
pass, though the tests testing for various optimizations
of course don't (they're marked as test_broken).

vchuravy committed:
Merge pull request #39336 from JuliaLang/jb/finalizerenabletest

fix failing `enable_finalizers` test

kimikage committed:
Fix performance regression in broadcasting with CartesianIndices (#39333)

* Fix performance regression in broadcasting with CartesianIndices

This avoids the boundary check due to a change in the implementation
of iteration using `CartecianIndices` in PR #37829.
This is a workaround on the caller side and does not change
the iteration mechanism itself.


Co-authored-by: Matt Bauman <mbauman@gmail.com>
Co-authored-by: thofma <thofma@gmail.com>

Keno committed:
Fix OC code emission (#39346)

Fixes issue currently seen on CI - master's configuration
changed since the OC PR ran, so it wasn't caught before
merging.

StefanKarpinski committed:
LibGit2: patch to pass hostkey & port to host verify callback (#39324)

It seems that no one actually verifies SSH host identity with libgit2
because the callback doesn't give enough information do so correctly:

- It doesn't give the actual host key fingerprint, but rather three
  different hashes thereof. This means we cannot distinguish a known
  hosts entry that has a different type (`ssh-rsa`, `ssh-dsa`, etc.)
  from an entry with a matching type and a fingerprint mismatch: the
  former should be treated as an unknown host whereas the latter is a
  host key mismatch; they cannot be distinguished without this patch.

- If the user connects on a non-default port (i.e. not 22), this is not
  passed to the callback in any way. Since there can be different known
  host entries for different ports and they should be treated as
  distinct, this also means the current API cannot be used to verify
  hosts serving SSH on non-standard ports. This patch passes the port.

I will try to upstream some version of this patch to libgit2. The same
patch has already been applied to the LibGit2 JLL.

Fixes #38777.
Might fix https://github.com/JuliaLang/Pkg.jl/issues/2334.

oxinabox committed:
define keys on RegexMatch (#37299)

baggepinnen committed:
Correct typo in module docs (#39337)

fonsp committed:
Docs: typo in Base.open (#39339)

KristofferC committed:
add a test that error can be inferred to bottom (#35806)

oxinabox committed:
Make RegexMatch iterable (#34355)

add iterate, length, and eltype to RegexMatch

Co-authored-by: Curtis Vogt <curtis.vogt@gmail.com>

vchuravy committed:
Merge pull request #39065 from fingolfin/mh/jl_new_task-stack-alloc

Fix jl_new_task to avoid allocating stack twice

DebadityaPal committed:
changed structure for exp, log and sqrt matrix operations

DebadityaPal committed:
Merge conflicts resolved

DebadityaPal commented:
@dkarrasch do you mind if I close this PR and start a fresh one? This branch fell way behind and all the commits got added to this.

dkarrasch commented:
You could try to rebase it onto current master. Here's what I do after pulling the current master to your local master:
```
git pull --rebase origin master
++ resolve merge conflicts ++
git push --force
```
If that doesn't work, sure, open a new PR.
